File: trees/226_invert_binary_tree.py
"""
Problem: Invert Binary Tree
LeetCode #: 226
Difficulty: Easy
Link: https://leetcode.com/problems/invert-binary-tree/

Approach: Recursive DFS — swap left and right children at every node.
Time Complexity: O(n)
Space Complexity: O(h) where h = height of tree
"""

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None

        root.left, root.right = root.right, root.left
        self.invertTree(root.left)
        self.invertTree(root.right)

        return root

File: trees/226_invert_binary_tree.java
/*
 * Problem: Invert Binary Tree
 * LeetCode #: 226
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/invert-binary-tree/
 *
 * Approach: Recursive depth-first traversal swapping left and right subtrees.
 * Time Complexity: O(n)
 * Space Complexity: O(h) where h is height of tree
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) { this.val = val; }
}

public class Solution {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) return null;
        TreeNode temp = root.left;
        root.left = invertTree(root.right);
        root.right = invertTree(temp);
        return root;
    }
}
